[SUREFIRE-2076] BufferOverflowException when encoding message with null runMode

zoltanmeze opened the pull request:
Per @Tibor17's comment https://github.com/apache/maven-surefire/pull/518#issuecomment-1099706229 and https://github.com/apache/maven-surefire/pull/518#issuecomment-1108371215 
and after some dicsussion (see comment after https://github.com/apache/maven-surefire/pull/518#issuecomment-1099706229) pulled this into a separate issue.

Jira ticket: https://issues.apache.org/jira/browse/SUREFIRE-2076

[AbstractStreamEncoder#encodeHeader](https://github.com/apache/maven-surefire/blob/959c1e9cabb8d06c72f5ebd7eb6e56e9987eccf8/surefire-api/src/main/java/org/apache/maven/surefire/api/stream/AbstractStreamEncoder.java#L86) stores runMode part in at least 3 bytes:
- 1 byte for length
- 1 byte delimiter
- runMode content in length-bytes
- 1 byte delimiter 

In case of null runMode the encoded part becomes `0::` (exactly 3 bytes length)

The issue is that [AbstractStreamEncoder#estimateBufferLength](https://github.com/apache/maven-surefire/blob/959c1e9cabb8d06c72f5ebd7eb6e56e9987eccf8/surefire-api/src/main/java/org/apache/maven/surefire/api/stream/AbstractStreamEncoder.java#L184) is not expecting/couting any bytes for runMode part in case of null runMode. This results in in BufferOverflowException becase the byte size of the entire message is underestimated.

Exception thrown:
```
java.nio.BufferOverflowException
	at java.nio.Buffer.nextPutIndex(Buffer.java:547)
	at java.nio.HeapByteBuffer.put(HeapByteBuffer.java:172)
	at org.apache.maven.surefire.api.stream.AbstractStreamEncoder.encodeString(AbstractStreamEncoder.java:127)
	at org.apache.maven.surefire.api.stream.AbstractStreamEncoder.encodeStringData(AbstractStreamEncoder.java:171)
	at org.apache.maven.surefire.api.stream.AbstractStreamEncoder.encode(AbstractStreamEncoder.java:157)
	at org.apache.maven.surefire.booter.spi.EventChannelEncoder.encodeMessage(EventChannelEncoder.java:398)
	at org.apache.maven.surefire.booter.spi.EventChannelEncoder.setOutErr(EventChannelEncoder.java:188)
	at org.apache.maven.surefire.booter.spi.EventChannelEncoder.testOutput(EventChannelEncoder.java:183)
	at org.apache.maven.surefire.api.booter.ForkingRunListener.writeTestOutput(ForkingRunListener.java:113)
	at org.apache.maven.surefire.api.booter.ForkingRunListener.writeTestOutput(ForkingRunListener.java:44)
	at org.apache.maven.surefire.common.junit4.JUnit4RunListener.writeTestOutput(JUnit4RunListener.java:235)
	at org.apache.maven.surefire.api.report.ConsoleOutputCapture$ForwardingPrintStream.println(ConsoleOutputCapture.java:144)
```


Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SUREFIRE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SUREFIRE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `SUREFIRE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean install`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Tibor17 commented:
After the #518 is merged, you have to rebase this PR. Then we will run the CI again.

Tibor17 commented:
#518 was merged, let's concentrate on this one.

zoltanmeze commented:
> After the #518 is merged, you have to rebase this PR. Then we will run the CI again.

Rebased onto latest master.

> #518 was merged, let's concentrate on this one.

The issue should be easily reproducible with  [AbstractStreamEncoderTest#testStdErrStreamEmptyMessageNullRunMode](https://github.com/apache/maven-surefire/pull/529/files#diff-d64bde0920c0b26eb5bec21735b15e8f57e2ca8b6fbd1b387f2f47ccf7e2d5d3R1255). Test implemented similar way as `testStdErrStreamEmptyMessageNullTestId` from [SUREFIRE-2056](https://issues.apache.org/jira/browse/SUREFIRE-2056).

Expected length also changed in [AbstractStreamEncoderTest#shouldComputeStreamPreemptiveLength
](https://github.com/apache/maven-surefire/pull/529/files#diff-1b8c3d63f48dfb2501c224494ef498d3613047545954ddfd7606f4163a7d8ad0R140) in cases where runMode is set to null, in those cases length is now greater by 3. Also included how the output should look like (without testId) in the comments.

Tibor17 reviewed (approved):
LGTM

Tibor17 commented:
@zoltanmeze This change looks fine. Pls check the method `estimateBufferLength()` completely if you see something we need to improve. thx

zoltanmeze commented:
> @zoltanmeze This change looks fine. Pls check the method `estimateBufferLength()` completely if you see something we need to improve. thx

Looks fine to me. Checked the usages a bit, was mostly looking if long and integer counts are properly set.

One thing I noticed, there is a null smartStackTrace included in encode on [EventChannelEncoder#L221](https://github.com/apache/maven-surefire/blob/master/surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/EventChannelEncoder.java#L221) but that same null string in not included in estimateBufferLength on [EventChannelEncoder.java#L217](https://github.com/apache/maven-surefire/blob/master/surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/EventChannelEncoder.java#L217). On master without these changes it's possible to get BufferOverflowException in [EventChannelEncoderTest#testConsoleError](https://github.com/apache/maven-surefire/blob/master/surefire-booter/src/test/java/org/apache/maven/surefire/booter/spi/EventChannelEncoderTest.java#L1018) by reducing message to only two characters or on this branch with empty string.

Tibor17 commented:
@zoltanmeze 
Exactly! The line should be `int bufferMaxLength = estimateBufferLength( BOOTERCODE_CONSOLE_ERROR.getOpcode().length(), null, encoder, 0, 0, message, null, stackTrace );` which includes `null` string. This should be fixed as well. Would you continue on this?

zoltanmeze commented:
Included missing null in [6e5a49c](https://github.com/apache/maven-surefire/pull/529/commits/6e5a49cb9fc8c6767a350d3bf530cd109354df4f). 

I don't think it's worth adding a unit test for this. This was obviously just missing. There are a couple of tests covering `consoleErrorLog`, but usually buffer length is overestimated because strings are allocating more space than what they actually need with `encoder.maxBytesPerChar() * s.length()`, so this was not causing any issues.

Tibor17 commented:
If there are no objections, I would proceed with this PR.

zoltanmeze commented:
> If there are no objections, I would proceed with this PR.

No objections from me

Tibor17 commented:
Thx for contributing.

jira-importer commented:
Resolve #3117
